lang/src/org/partiql/lang/eval/visitors/PartiqlAstSanityValidator.kt

```diff
@@ -0,0 +1,105 @@
+
+/*
+ * Copyright 2019 Amazon.com, Inc. or its affiliates.  All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ *  You may not use this file except in compliance with the License.
+ * A copy of the License is located at:
+ *
+ *      http://aws.amazon.com/apache2.0/
+ *
+ *  or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific
+ *  language governing permissions and limitations under the License.
+ */
+
+package org.partiql.lang.eval.visitors
+
+import com.amazon.ionelement.api.IntElement
+import com.amazon.ionelement.api.IntElementSize
+import org.partiql.lang.ast.IsCountStarMeta
+import org.partiql.lang.ast.passes.SemanticException
+import org.partiql.lang.domains.PartiqlAst
+import org.partiql.lang.domains.addSourceLocation
+import org.partiql.lang.domains.errorContextFrom
+import org.partiql.lang.errors.ErrorCode
+import org.partiql.lang.errors.Property
+import org.partiql.lang.errors.PropertyValueMap
+import org.partiql.lang.eval.err
+import org.partiql.lang.eval.errIntOverflow
+
+/**
+ * Provides rules for basic AST sanity checks that should be performed before any attempt at further AST processing.
+ * This is provided as a distinct visitor so that all other visitors may assume that the AST at least passed the
+ * checking performed here.
+ *
+ * Any exception thrown by this class should always be considered an indication of a bug in one of the following places:
+ *
+ * - [org.partiql.lang.syntax.SqlParser]
+ * - A visitor transform pass (internal or external)
+ *
+ */
+object PartiqlAstSanityValidator {
```

Comment:
This should be implementing, `PartiqlAst.Visitor`.

What you have here doesn't catch errors in child nodes.  

Please add some tests which with bad non-root nodes.

Reply:
Good catch. Should be fixed in the newest revision. Also added new subquery tests. Thanks